### PR TITLE
`wac plug` with registry support

### PR DESCRIPTION
**Pull request by calvinrp**
Updates Warg to `0.6.0` release with support for federating across multiple registries.

Adds `wac plug` registry support (feature flag still but a default one).

Existing behavior is unchanged:
`wac plug my-component.wasm --plug helpful.wasm -o plugged.wasm`

But now, you can mix in registry package names or local file paths. In any combination. Kind of fun.

It determines whether to look in the registry or local path by trying to parse as valid `namespace:package-name`. Also, parses with versions in `namespace:package-name@0.1.0` syntax. If it is a valid package name, then it looks in the registry otherwise tries local file paths. So `path/some-file.wasm` would be unambiguously a local file path.

`wac plug macovedj:hashimap-service --plug macovedj:my-component@0.0.1 -o plugged.wasm`
(those are [packages](https://wa.dev/macovedj) from @macovedj WasmIO demo and that command will work if Warg is configured)

or

`wac plug hashimap-service.wasm --plug macovedj:my-component@0.0.1 -o plugged.wasm`

### CODE REVIEW

**Review by macovedj — APPROVED** [review]

**Comment by calvinrp** [comment]
@macovedj thanks for the review. I definitely want the other maintainers to weigh in on this.

File: crates/wac-resolver/src/lib.rs [hunk 1]
```diff
@@ -56,22 +65,22 @@ pub enum Error {
         #[label(primary, "package `{name}` does not exist")]
         span: SourceSpan,
     },
-    /// The requested package version has been yanked.
+    /// The requested package version has been yanked or does not exist.
     #[cfg(feature = "registry")]
-    #[error("version {version} of package `{name}` has been yanked")]
-    PackageVersionYanked {
+    #[error("version {version} of package `{name}` has been yanked or does not exist")]
+    PackageVersionYankedOrDoesNotExist {
```
Comment (fibonacci1729) [inline crates/wac-resolver/src/lib.rs:71]: Any reason to not have a separate variant here for `PackageVersionDoesNotExist` (or add an `Option<semver::Version>` to the `PackageDoesNotExist` variant)?

**Review by fibonacci1729 — COMMENTED** [review]
This looks good to me; I'd like to give @rylev a chance to review when he returns next week.

File: crates/wac-resolver/src/lib.rs (hunk 1 above)
Comment (calvinrp) [reply crates/wac-resolver/src/lib.rs:71]: At the moment, `download_exact()` returns an error that does distinguish between not found and yanked. That should be improved on the Warg side.

https://github.com/bytecodealliance/registry/blob/main/crates/client/src/lib.rs#L700-L713

Comment (fibonacci1729) [reply crates/wac-resolver/src/lib.rs:71]: In that case, should we just call this `PackageVersionDoesNotExist` until we can distinguish?

Comment (calvinrp) [reply crates/wac-resolver/src/lib.rs]: Done

File: crates/wac-resolver/src/registry.rs [hunk 2]
```diff
@@ -82,242 +68,120 @@ impl RegistryPackageResolver {
...
+                        }
+                    })?,
+                    (key.version.cloned(), *span),
+                ))
+            })
+            .collect::<Result<IndexMap<PackageName, (Option<Version>, SourceSpan)>, Error>>()?;
+
+        // fetch required package logs and return error if any not found
+        if let Some(bar) = self.bar.as_ref() {
+            bar.init(1);
```
Comment (rylev) [inline crates/wac-resolver/src/registry.rs:89]: I'm not familiar with how the progress bar works, but I'm surprised to see this hardcoded to `1`. Do we need a progress bar if there's only one thing to do?

File: src/commands/plug.rs [hunk 3]
```diff
@@ -53,24 +151,66 @@ impl PlugCommand {
...
+                    PackageRef::RegistryPackage((name, version)) => {
+                        let client = client.as_ref().ok_or_else(|| {
+                            anyhow::anyhow!(
+                                "Warg registry is not configured. Package `{name}` was not found."
+                            )
+                        })?;
+
+                        let path = if let Some(ver) = version {
+                            let download = client.download_exact(name, ver).await?;
+                            println!(
```
Comment (rylev) [inline src/commands/plug.rs:181]: We're not currently printing anything else when running the plug command. I think all of these `println!`s should be `log::debug` statements instead. In a follow up, we can figure out which output to always show the user.

**Review by rylev — CHANGES_REQUESTED** [review]
Looks good to me. I just would like to see the comment about the usage of `println!` in the plug command addressed.

File: crates/wac-resolver/src/registry.rs (hunk 2 above)
Comment (calvinrp) [reply crates/wac-resolver/src/registry.rs]: Fixed. It was left over from the previous behavior.

File: src/commands/plug.rs (hunk 3 above)
Comment (calvinrp) [reply src/commands/plug.rs]: Fixed.

**Review by rylev — APPROVED** [review]

**Pull request merged**